### ISSUE TITLE
Compatibility issues with kernel 4.4+

### ISSUE THREAD

**Issue by SketchStick** [DIGImend/digimend-kernel-drivers#31]
I have a Yiynova MVP22U (V4) and I'm running Arch with all the latest packages.

Using the latest git with 4.5.1, I was finding my tablet not listed in xinput and not responding at all. In some circumstances I had it listed, but the identifiers and a whole bunch of features were missing and the lack of sub-pixel precision was making any strokes jittery.

I tried a wide range of kernel versions and found that the 4.4 release was when it broke, 4.3.3 still working perfectly and being my fall-back for now. To make sure this wasn't an Arch issue I also tested this on Debian in a VM and had the same result...

**Comment by spbnick** [comment]
Thanks for the report! I'll try to look at this this week.

**Comment by spbnick** [comment]
First of all, I see that it simply doesn't build on newer kernels, because the HID driver interface has changed slightly in v4.4 (to be fixed). So I suspect that you simply didn't have the driver installed for the newer version. How did you verify it was there?

**Comment by spbnick** [comment]
I mean I'll be fixing my driver to use the updated HID driver interface in the kernel.

**Comment by SketchStick** [comment]
Well actually over here it appears to build, at least make didn't throw any errors and the module appears to load.
http://pastebin.com/raw/t3GWxwAS

**Comment by SketchStick** [comment]
Actually the issues I'm having with the in-kernel version could just be due to the lack of Yiynova support... Here's the xinput: http://pastebin.com/raw/A3vdz8DX

**Comment by spbnick** [comment]
Yes, the difference is due to the in-kernel driver being used. Do you have dmesg output for when you plug in the tablet on the broken kernel?

**Comment by SketchStick** [comment]
Here's the dmesg when plugging it in, 4.4.8 with module loaded: http://pastebin.com/raw/4VHad8fQ

**Comment by spbnick** [comment]
Still, since the module builds (even with a warning), it should install and load. I'm not sure what would exactly happen when that module is used, but I suspect it should either work fine, or completely fail to initialize. I'm not sure if another (generic) driver will be retried on failure (doesn't look like it).

**Comment by spbnick** [comment]
The dmesg output you provide shows the generic HID driver loading (which can't handle the tablet properly). So it either doesn't install, or is not rebound by hid-rebind (which might be a separate issue).

**Comment by SketchStick** [comment]
Well here's the dmesg output for the others, just for the hell of it: http://pastebin.com/raw/C99UVLDk

**Comment by spbnick** [comment]
Yeah, the older kernel output shows the tablet being rebound to the out-of-tree driver. So, if the driver indeed installed with the newer kernel, then rebinding of the module is going wrong.

**Comment by spbnick** [comment]
I'm still suspicious why the module hasn't loaded on the newer kernel, but please try this patch which should fix the compatibility:

https://gist.github.com/spbnick/a7ec56164bc830ff6f6bc56281c60e18

Meanwhile I'll work on making the driver build properly on both the older and the newer kernels.

**Comment by SketchStick** [comment]
Had to manually apply the patch but it's all working! 4.4.8 with the module loaded, shows up in xinput and draws nice clean strokes.

**Comment by spbnick** [comment]
Nice, thanks for testing!

**Comment by SketchStick** [comment]
Thank you for the wonderful driver!

I was also wondering if you had any plans to push a new version to the kernel? It seems like you've added quite a few devices and it'd be nice to have those working out of the box.

**Comment by spbnick** [comment]
Yes, I have a plan to contribute the recent changes to the kernel, but not until I finish my Finnish courses which take almost all of my free time. That can happen in about a month.

**Comment by SketchStick** [comment]
Just a final confirmation that the latest commit works perfectly under 4.5.1! All the best with your courses and thanks again for everything.

**Comment by spbnick** [comment]
You're welcome :)
